ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp

```diff
@@ -24,231 +23,230 @@
+      case BATTERY_LEVEL_FULL:
+        return gp_full;
+      default:
+        break;
     }
+  }
+
+  return 0;
 }
 
 float gamepad_get_button_threshold(int device) {
-	
+
 }
-	
+
 void gamepad_set_button_threshold(int device, float threshold) {
 
 }
-	
-void gamepad_set_vibration(int device, float left, float right) {
-	XINPUT_STATE controllerState;
-    // Create a Vibraton State
-    XINPUT_VIBRATION vibration;
 
-    // Zeroise the Vibration
-    ZeroMemory(&vibration, sizeof(XINPUT_VIBRATION));
+void gamepad_set_axis_deadzone(int device, float deadzone) {
+
+}
+
+void gamepad_set_vibration(int device, float left, float right) {
+  XINPUT_VIBRATION vibration = {};
 
-    // Set the Vibration Values
-    vibration.wLeftMotorSpeed = left * 65535;
-    vibration.wRightMotorSpeed = right * 65535;
+  vibration.wLeftMotorSpeed = left * 65535;
+  vibration.wRightMotorSpeed = right * 65535;
 
-    // Vibrate the controller
-    XInputSetState(device, &vibration);
+  XInputSetState(device, &vibration);
 }
-	
+
 int gamepad_axis_count(int device) {
-	return 2;
+  return 4;
 }
-	
+
 float gamepad_axis_value(int device, int axis) {
-	XINPUT_STATE controllerState;
-    // Zeroise the state
-    ZeroMemory(&controllerState, sizeof(XINPUT_STATE));
-
-    // Get the state
-    XInputGetState(device, &controllerState);
-	float ret;
-	switch (axis) {
-		case gp_axislh:
-			ret = controllerState.Gamepad.sThumbLX;
-			break;
-		case gp_axislv:
-			ret = controllerState.Gamepad.sThumbLY;
-			break;
-		case gp_axisrh:
-			ret = controllerState.Gamepad.sThumbRX;
-			break;
-		case gp_axisrv:
-			ret = controllerState.Gamepad.sThumbRY;
-			break;
-		default:
-			return -1;
-	}
-	if (ret > 0) {
-		return ret / 32767;
-	} else {
-		return ret / 32768;
-	}
-}
-
-WORD digitalButtons[20] = { XINPUT_GAMEPAD_A, XINPUT_GAMEPAD_X, XINPUT_GAMEPAD_B, XINPUT_GAMEPAD_Y, XINPUT_GAMEPAD_LEFT_SHOULDER, 
-	XINPUT_GAMEPAD_RIGHT_SHOULDER, 0, 0, XINPUT_GAMEPAD_BACK, XINPUT_GAMEPAD_START, XINPUT_GAMEPAD_LEFT_THUMB, XINPUT_GAMEPAD_RIGHT_THUMB, 
-	XINPUT_GAMEPAD_DPAD_UP, XINPUT_GAMEPAD_DPAD_DOWN, XINPUT_GAMEPAD_DPAD_LEFT, XINPUT_GAMEPAD_DPAD_RIGHT, 0, 0, 0, 0 };
-	
+  XINPUT_STATE controllerState = {};
+
+  DWORD dwResult = XInputGetState(device, &controllerState);
+
+  if (dwResult == ERROR_SUCCESS) {
+    float ret = 0;
+    switch (axis) {
+      case gp_axislh:
+          ret = controllerState.Gamepad.sThumbLX;
+        break;
+      case gp_axislv:
+          ret = -controllerState.Gamepad.sThumbLY - 1;
+        break;
+      case gp_axisrh:
+          ret = controllerState.Gamepad.sThumbRX;
+        break;
+      case gp_axisrv:
+          ret = -controllerState.Gamepad.sThumbRY - 1;
+        break;
+      default:
+        return 0;
+    }
+    if (ret > 0) {
+      return ret / 32767;
+    } else {
+      return ret / 32768;
+    }
+  } else {
+    return 0;
+  }
+}
+
+WORD digitalButtons[20] = {
+    XINPUT_GAMEPAD_A, XINPUT_GAMEPAD_X, XINPUT_GAMEPAD_B, XINPUT_GAMEPAD_Y,
```

Comment:
XINPUT_GAMEPAD_X and XINPUT_GAMEPAD_B need to be switched. gp_face2 is B, and gp_face3 is X.

Reply:
I took care of it in my updated XInput pull request that I am creating: https://github.com/enigma-dev/enigma-dev/commit/4d4f11f616ac682679f2213affb6453c3da3c02c

You were correct, thanks for the help.